oak_functions/loader/src/server.rs

```diff
@@ -0,0 +1,121 @@
+//
+// Copyright 2021 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use anyhow::Context;
+use http::{request::Request, response::Response};
+use hyper::{
+    service::{make_service_fn, service_fn},
+    Body, Server, StatusCode,
+};
+use log::info;
+use std::{net::SocketAddr, sync::Arc};
+
+const MAIN_FUNCTION_NAME: &str = "main";
+
+// An ephemeral request handler with a Wasm module for handling the requests.
+#[derive(Clone)]
+pub(crate) struct WasmHandler {
+    // Wasm module to be served on each invocation.
+    module: Arc<wasmi::Module>,
+}
+
+impl WasmHandler {
+    pub(crate) fn new(wasm_module_bytes: &[u8]) -> anyhow::Result<Self> {
+        let module = wasmi::Module::from_buffer(&wasm_module_bytes)?;
+        Ok(WasmHandler {
+            module: Arc::new(module),
+        })
+    }
+    pub(crate) fn handle_request(&self, req: &Request<Body>) -> anyhow::Result<Response<Body>> {
+        // TODO(#1919): Make request available to the Wasm module via ABI functions.
+        info!("The request is: {:?}", req);
+
+        let instance = wasmi::ModuleInstance::new(&self.module, &wasmi::ImportsBuilder::default())
+            .context("failed to instantiate Wasm module")?
+            .assert_no_start();
+
+        let result = instance.invoke_export(MAIN_FUNCTION_NAME, &[], &mut wasmi::NopExternals);
+
+        info!(
+            "{:?}: Running Wasm module completed with result: {:?}",
+            std::thread::current().id(),
+            result
+        );
+
+        // TODO(#1919): Get the actual response from the Wasm module via ABI functions.
+        http::response::Builder::new()
+            .status(StatusCode::BAD_REQUEST)
+            .body(Body::from("Not implemented yet.\n"))
+            .context("Couldn't create response")
+    }
+}
+
+/// An HTTP server that serves the `main` function of a Wasm module on each invocation.
+pub struct WasmServer {
+    /// Server address to listen for client requests on.
+    address: SocketAddr,
+    /// The Wasm handler that runs the given Wasm module per incoming request.
+    wasm_handler: WasmHandler,
+}
+
+impl WasmServer {
+    /// Creates a [`WasmServer`] instance listening on the given port, serving the given Wasm
```

Comment:
I don't think this actually starts listening on the given port / address, though maybe `create` and `start` could be combined in one (up to you); or is there anything useful that a non-started `WasmServer` can be used for?